### PR TITLE
CPP-499 allow local bind

### PR DESCRIPTION
**Pull request by kw217**
Fix https://datastax-oss.atlassian.net/browse/CPP-499, i.e., add `cass_cluster_set_local_address` to allow clients to bind a local IP address before connecting to Cassandra. This is important for hosts using multiple network interfaces with strict network separation requirements.

### CODE REVIEW

**Comment by kw217** [comment]
CLA now signed.

**Comment by mikefero** [comment]
@kw217 It appears the CLA has been delivered to your inbox, but has not been completely signed. Can you perform the CLA signing again using the DocuSign tool ensuring you click finish after completing the signing process?

Once that is completed the PR will be updated automatically and the label will be removed.

**Comment by kw217** [comment]
Clicking on the link in the email gives me the DocuSign site, and it says on the top of the document "This document is now complete." Would you like me to download the PDF from DocuSign and send it to you? Because of the wording of the CLA and because the work is being done on company time, my company legal department required someone in my line to sign it on my behalf, which was permitted by DocuSign and which they have done. Could it be that the bot is confused by this?

FTR: the summary events from DocuSign are:
```
Envelope Sent Hashed/Encrypted 4/18/2018 5:36:57 AM
Certified Delivered Security Checked 4/18/2018 6:02:35 AM
Signing Complete Security Checked 4/18/2018 6:09:01 AM
Completed Security Checked 4/18/2018 6:09:01 AM
```

File: gtests/src/integration/tests/test_control_connection.cpp [hunk 1]
```diff
@@ -22,6 +22,17 @@
...
  */
 class ControlConnectionTests : public Integration {
+
+public:
+
+  void SetUp() {
+    // Call the parent setup function (don't automatically start session,
+    // because we don't want any connections established until we have
+    // set up the cluster).
+    is_session_requested_ = false;
```
[the remark this answers is not in the document]
Comment (mikefero) [reply gtests/src/integration/tests/test_control_connection.cpp]: This is actually a nice improvement as almost all of the tests create a cluster and session for themselves, meaning their are currently multiple cluster and sessions for each test. The only test that I see failing would be [**FullOutage**](https://github.com/kw217/cpp-driver/blob/a8ac0b496637adb524b9b997bb45443b3b6bd20e/gtests/src/integration/tests/test_control_connection.cpp#L569-L611); before the [**cluster is stopped**](https://github.com/kw217/cpp-driver/blob/a8ac0b496637adb524b9b997bb45443b3b6bd20e/gtests/src/integration/tests/test_control_connection.cpp#L590) a call to `connect()` should be made to establish the `session_` instance in the base class.

File: include/cassandra.h [hunk 2]
```diff
@@ -946,6 +946,39 @@ CASS_EXPORT CassError
...
+ * @public @memberof CassCluster
+ *
+ * @param[in] cluster
+ * @param[in] name IP address to bind, or empty string for no binding.
+ * Only numeric addresses are supported; no resolution is done.
+ * @return CASS_OK if successful, otherwise an error occurred.
+ */
+CASS_EXPORT CassError
+cass_cluster_set_local_address(CassCluster* cluster,
+                      const char* name);
```
[the remark this answers is not in the document]
Comment (kw217) [reply include/cassandra.h]: tnx, fixed.

File: include/cassandra.h [hunk 3]
```diff
@@ -946,6 +946,39 @@ CASS_EXPORT CassError
...
+ *
+ * @param[in] cluster
+ * @param[in] name
+ * @param[in] name_length
+ * @return same as cass_cluster_set_local_address()
+ *
+ * @see cass_cluster_set_local_address()
+ */
+CASS_EXPORT CassError
+cass_cluster_set_local_address_n(CassCluster* cluster,
```
[the remark this answers is not in the document]
Comment (kw217) [reply include/cassandra.h]: tnx, fixed.

File: src/connection.cpp [hunk 4]
```diff
@@ -244,9 +245,20 @@ Connection::Connection(uv_loop_t* loop,
...
 
-  SslContext* ssl_context = config_.ssl_context();
-  if (ssl_context != NULL) {
-    ssl_session_.reset(ssl_context->create_session(host));
+  const Address* local_address = config_.local_address();
+  if (local_address) {
+    int rc = uv_tcp_bind(&socket_, local_address->addr(), 0);
+    if (rc) {
+      ok = false;
+      notify_error("Unable to bind local address: " + std::string(UV_ERRSTR(rc, loop_)));
```
[the remark this answers is not in the document]
Comment (kw217) [reply src/connection.cpp]: Ah - not sure what I was thinking! Fixed, thanks.

File: gtests/src/integration/tests/test_control_connection.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kw217) [reply gtests/src/integration/tests/test_control_connection.cpp]: OK, I've added that in.

**Pull request merged**